libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java

```diff
@@ -177,25 +179,34 @@ public boolean isValidRouteProfile(String routeProfile) {
    */
   @Nullable
   public BannerInstructions findCurrentBannerInstructions(LegStep currentStep, double stepDistanceRemaining) {
-    if (isValidStep(currentStep) && hasInstructions(currentStep.bannerInstructions())) {
-      int roundedDistanceRemaining = (int) stepDistanceRemaining;
-      List<BannerInstructions> instructions = new ArrayList<>(currentStep.bannerInstructions());
-      Iterator<BannerInstructions> instructionsIterator = instructions.iterator();
-      while (instructionsIterator.hasNext()) {
-        BannerInstructions instruction = instructionsIterator.next();
+    if (isValidBannerInstructions(currentStep)) {
+      List<BannerInstructions> instructions = getBannerInstructions(currentStep);
+      for (BannerInstructions instruction : instructions) {
         double distanceAlongGeometry = instruction.distanceAlongGeometry();
-        if (distanceAlongGeometry < roundedDistanceRemaining) {
-          instructionsIterator.remove();
+        if (distanceAlongGeometry >= stepDistanceRemaining) {
+          return instruction;
         }
       }
-      int instructionIndex = checkValidIndex(instructions);
-      if (instructions.size() > ZERO_INSTRUCTIONS) {
-        return instructions.get(instructionIndex);
-      }
+      return instructions.get(FIRST_INSTRUCTION);
     }
     return null;
   }
 
+  private boolean isValidBannerInstructions(LegStep currentStep) {
+    return isValidStep(currentStep) && hasInstructions(currentStep.bannerInstructions());
+  }
+
+  private List<BannerInstructions> getBannerInstructions(LegStep currentStep) {
```

Comment:
Here is fine to pass in `List<BannerInstructions>` as you suggested in the first place. No need to pass in the whole `LegStep` object if we're only interested in the banner instructions list.